crates/rpc/src/jsonrpc/websocket/logic.rs

```diff
@@ -315,6 +346,76 @@ async fn header_subscription(
     }
 }
 
+async fn event_subscription(
+    msg_sender: mpsc::Sender<ResponseEvent>,
+    mut events: broadcast::Receiver<Vec<Vec<Event>>>,
+    subscription_id: u32,
+    filter: EventFilterParams,
+) {
+    use broadcast::error::RecvError;
+    let key_filter_is_empty = filter.keys.iter().flatten().count() == 0;
+    let keys: Vec<std::collections::HashSet<_>> = filter
+        .keys
+        .iter()
+        .map(|keys| keys.iter().collect())
+        .collect();
+    'outer: loop {
+        match events.recv().await {
+            Ok(events) => {
+                let events = events.into_iter().flat_map(|events| {
+                    events
+                        .into_iter()
+                        .filter(|event| match filter.address {
+                            Some(address) => event.from_address == address,
+                            None => true,
+                        })
+                        .filter(|event| {
+                            if key_filter_is_empty {
+                                return true;
+                            }
+
+                            if event.keys.len() < keys.len() {
+                                return false;
+                            }
+
+                            event
+                                .keys
+                                .iter()
+                                .zip(keys.iter())
+                                .all(|(key, filter)| filter.is_empty() || filter.contains(key))
+                        })
+                });
+                for event in events {
+                    let response = ResponseEvent::Event(SubscriptionItem {
+                        subscription_id,
+                        item: Arc::new(event),
+                    });
+                    if msg_sender.send(response).await.is_err() {
+                        break 'outer;
+                    }
+                }
+            }
+            Err(RecvError::Closed) => break,
+            Err(RecvError::Lagged(amount)) => {
+                tracing::info!(
+                    amount,
+                    "Lagging header stream, missed some events, closing subscription"
```

Comment:
```suggestion
                    "Lagging events stream, missed some events, closing subscription"
```
Though I wonder if we should change the wording to something like
```
tracing::debug!(
      %subscription_id,
      kind="event",
      "Subscription consumer too slow, closing."
);
```
and something similar for the error response.